macros/compare_column_values_verbose.sql

```diff
@@ -0,0 +1,32 @@
+{% macro compare_column_values_verbose(a_query, b_query, primary_key, column_to_compare) -%}
+  {{ return(adapter.dispatch('compare_column_values_verbose', 'audit_helper')(a_query, b_query, primary_key, column_to_compare)) }}
+{%- endmacro %}
+
+
+{% macro default__compare_column_values_verbose(a_query, b_query, primary_key, column_to_compare) -%}
+with a_query as (
+    {{ a_query }}
+),
+
+b_query as (
+    {{ b_query }}
+)
+    select
+        coalesce(a_query.{{ primary_key }}, b_query.{{ primary_key }}) as {{ primary_key }},
+        '{{ column_to_compare }}' as column_name,
+        a_query.{{ column_to_compare }} = b_query.{{ column_to_compare }} as perfect_match,
+        a_query.{{ column_to_compare }} is null as null_in_a,
+        b_query.{{ column_to_compare }} is null as null_in_b,
+        a_query.{{ primary_key }} is null as missing_from_a,
+        b_query.{{ primary_key }} is null as missing_from_b,
+        a_query.{{ column_to_compare }} != b_query.{{ column_to_compare }} and
+            (a_query.{{ column_to_compare }} is not null or b_query.{{ column_to_compare }} is not null)
+          as conflicting_values
+           -- considered a conflict if the values do not match AND at least one of the values is not null.
+
+    from a_query
+
+    full outer join b_query on a_query.{{ primary_key }} = b_query.{{ primary_key }}
```

Comment:
@joellabes I hate to open a can of worms, but would we want to add `or (a_query.{{ primary_key }} is null and b_query.{{ primary_key }} is null)` to [this](https://github.com/dbt-labs/dbt-audit-helper/blob/dd5d2ed7933fc5a82dc3dcc83ea7042d0222355a/macros/compare_column_values.sql#L42) as well? It's late, so I might be tired/missing a reason why we wouldn't want the same thing here!

Reply:
Yes I think you're right!! Good find

Reply:
@joellabes I'm having second thoughts about this. The primary key should never be null in either table. I feel like that's a reasonable expectation, and there should be a separate `not_null` test to catch data errors of that nature. Moreover, imagine if there were 10 records in one table where the primary key is null, and 10 records in another table. Since it's a full outer join, I think that would create fan-out. Right?

The point of this join (and test) is supposed to be to catch instances where a given primary key is missing in one of the tables (which would indicate a missing record) _or_ if the primary key matches, but the data conflicts. If the table has null values in the primary key column, that's an issue of a different nature.

Thoughts?

Reply:
> I'm having second thoughts about this.

Is "this" just the compare_column_values one, or both? If the former, then that's ok! if the latter, that's also OK but then you should probably remove the columns that talk about null PKs right?

Reply:
I meant both. When you say:

> you should probably remove the columns that talk about null PKs right

are you referring to these columns, `missing_from_a` and `missing_from_b`?

```
        a_query.{{ primary_key }} is null as missing_from_a,
        b_query.{{ primary_key }} is null as missing_from_b,
```

If so, I don't think they should be removed, because what they represent are columns where the PK is present in one query, but not the other. That doesn't mean either query has a null value anywhere in the PK column.

If you're referring to _these_ two columns, `null_in_a` and `null_in_b` (which I don't think you are, but just in case):
```
        a_query.{{ column_to_compare }} is null as null_in_a,
        b_query.{{ column_to_compare }} is null as null_in_b,
```

they represent rows where the PK is present in both queries, but the value in the column of interest is null.

So, in none of thee cases will there be a null PK in `a_query` or `b_query`; rather, the null values occur only as a result of a full outer join where both queries don't contain the same PKs.

Does that make sense? Admittedly this is all fairly confusing for me.

Reply:
🤷 I am also getting a bit lost. I think you should make a test seed that exercises all the different types of data you expect to encounter with varying null/not null states and validate that it works the way you think it should. As long as that works out then I'm happy